add __str__ to Expr classes

DavyCats opened the pull request:
I was working on a script to compile user documentation for WDL workflows and I found myself writing functions to render WDL expressions so the default values of inputs could be shown in these docs. It seemed to me that it might be be more appropriate to add these functions to the WDL bindings, hence this PR.

If this PR were to get merged, calling `str()` on most WDL.Expr classes would return a human-readable expression. The returned strings should look similar to how the expression would be written in WDL in most cases. For structs they currently would look like this, though: `(field: "A",  another_field: 1.1, yet_another_field: (another_struct: true))`.

Let me know what you think and if this is even something you're interested in. If so, could you point me to where I should add tests for this?

mlin commented:
Thanks for this! Eventually, we'd definitely like miniwdl to support writing back WDL source code (#22), perhaps even a [gofmt](https://golang.org/cmd/gofmt/) or [black](https://github.com/psf/black) type of tool one day. It's entirely missing right now, and this is a nice first step.

The current Travis test failure happened (ironically) because it wants the code formatted by black. Can you please try `make pretty` and commit the reformatted code to the branch? 

After that, we'll probably get a code coverage complaint. I would just add a few straightforward tests for this in [test_0eval.py](https://github.com/chanzuckerberg/miniwdl/blob/master/tests/test_0eval.py). A cute way to do these might be to `WDL.parse_expr()` a string constant, and then check if the `WDL.Expr` object stringifies back to the original string exactly. You can quickly go `python3 -m unittest -f tests/test_0eval.py` to avoid having to run the whole test suite (which takes like 10 minutes).

DavyCats commented:
I added the test and ran `make pretty` and it seems to be passing the tests now. I also implemented a somewhat smarter method of adding parentheses. Instead of just adding them everywhere it now checks if it is necessary based on the precedence specified in the WDL specs.

One thing I did run into was the struct literals. Since these are not super clearly defined in WDL 1.0 and they are likely changing in whatever the next version of WDL will be, I opted to return a Map literal instead.

Having a "WDLfmt" would be nice, currently we (at BioWDL) are using some custom [style guidelines](https://biowdl.github.io/styleGuidelines.html), but since we have to check it by hand during reviews, mistakes are likely to slip through. Having more uniformity between different projects/institutes would be good.

mlin commented on WDL/Expr.py:
```diff
@@ -677,6 +729,9 @@ def __init__(self, pos: SourcePosition, parts: List[str]) -> None:
         self.referee = None
 
+    def __str__(self):
+        return self.name
```
```suggestion
        return ".".join(self.namespace + [self.name])
```
